### PR TITLE
Add Push Notification example

### PR DESCRIPTION
**Pull request by nakajima**
Add example that uses https://github.com/xmtp/example-notification-server-go to receive push notifications.

### CODE REVIEW

**Comment by nakajima** [comment]
I wonder if _all_ of this should be in the example app actually instead of adding certain parts to the SDK. It feels like it'd be nice to have more batteries included for the SDK but also maybe this is too presumptive?

**Review by nplasterer — APPROVED** [review]
I think this seems fine to include. But I'll lean on @elisealix22 for her thoughts since there maybe be better places to show this as an example without making this base example too complicated.

**Review by elisealix22 — APPROVED** [review]
The push abstraction makes sense to me! I like that the persistence layer is still in the example app for now too. I ended up following these two examples for storing keys in the 1P app so keeping that flexible makes sense:
1. https://developer.apple.com/documentation/cryptokit/storing_cryptokit_keys_in_the_keychain
2. https://johncodeos.com/how-to-save-sensitive-data-in-keychain/

File: Sources/XMTP/Push/XMTPPush.swift [hunk 1]
```diff
@@ -0,0 +1,90 @@
+//
```
Comment (neekolas) [inline Sources/XMTP/Push/XMTPPush.swift:1]: What are the implications of having this live outside of the example folder?

It seems like if we want to treat this as a library, we are going to have to support it as such (documentation, API stability)

I wonder if it makes sense to move this into the example folder for now to lower expectations. Genuinely unsure of the right answer.

I'm not enough of a Swift expert to understand how much work that's going to put on the user to copy/paste/regenerate.

File: XMTPiOSExample/XMTPiOSExample/Views/ConversationListView.swift [hunk 2]
```diff
@@ -66,10 +69,29 @@ struct ConversationListView: View {
...
+			await add(conversations: conversations)
 		} catch {
 			print("Error loading conversations: \(error)")
 		}
 	}
+
+	func add(conversations: [Conversation]) async {
+		// Ensure we're subscribed to push notifications on these conversations
+		do {
+			try await XMTPPush.shared.subscribe(topics: conversations.map(\.topic))
```
Comment (neekolas) [inline XMTPiOSExample/XMTPiOSExample/Views/ConversationListView.swift:82]: For a full featured push notification example, we are going to have to subscribe to more than just conversation topics.

We are going to want to subscribe to the `invite` and `intro` topics to trigger a background update to the conversation list when a new invite/intro is received.

File: Sources/XMTP/Push/XMTPPush.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+			request.deliveryMechanism = Notifications_V1_DeliveryMechanism.with { delivery in
+				delivery.apnsDeviceToken = token
+				delivery.deliveryMechanismType = .apnsDeviceToken(token)
+			}
+		}
+
+		_ = await client.registerInstallation(request: request)
+	}
+
+	public func subscribe(topics: [String]) async throws {
```
Comment (neekolas) [inline Sources/XMTP/Push/XMTPPush.swift:67]: If we did want this to function more as a library that we expect other devs to use directly, I would expect this to include a `revoke` method that calls `deleteInstallation` as well as a way to `unsubscribe`.

Although for right now, neither of those are really needed to build a basic push example.

File: Sources/XMTP/Push/XMTPPush.swift (hunk 1 above)
Comment (nakajima) [reply Sources/XMTP/Push/XMTPPush.swift:1]: Having it over here means it's part of the SDK and anyone can just use it.

> I wonder if it makes sense to move this into the example folder for now to lower expectations. Genuinely unsure of the right answer.

Ya this is the biggest I have as well. I really like the idea that adding push to your app is as simple as using `XMTPPush`, but it does make a hard assumption that you're running [xmtp/example-notification-server-go](https://github.com/xmtp/example-notification-server-go) somewhere.

I also like the flexibility that anyone can write any software that subscribes to the network and delivers pushes in any format they want, I just wonder if that's a lot to be asking for folks who just want to get something working, especially iOS devs who may not really know much about setting up servers.

File: XMTPiOSExample/XMTPiOSExample/Views/ConversationListView.swift (hunk 2 above)
Comment (nakajima) [reply XMTPiOSExample/XMTPiOSExample/Views/ConversationListView.swift:82]: That's a good call, I should add it to this PR.

File: Sources/XMTP/Push/XMTPPush.swift (hunk 3 above)
Comment (nakajima) [reply Sources/XMTP/Push/XMTPPush.swift:67]: Since I'm leaning more toward the library approach, I think this would be good to add.

**Comment by nakajima** [comment]
Gonna merge this as is just because it's been open for a while. We can explore more of the line between library vs example in follow ups.

**Comment by neekolas** [comment]
👍🏼

**Pull request merged**